crates/task-impls/src/quorum_vote/drb_computations.rs

```diff
@@ -0,0 +1,126 @@
+use std::collections::{btree_map, BTreeMap};
+
+use hotshot_types::{
+    drb::{compute_drb_result, DrbResult, DrbSeedInput},
+    traits::node_implementation::{ConsensusTime, NodeType},
+};
+use tokio::{spawn, task::JoinHandle};
+
+/// Number of previous results and seeds to keep
+pub const KEEP_PREVIOUS_RESULT_COUNT: u64 = 8;
+
+/// Helper struct to track state of DRB computations
+pub struct DrbComputations<TYPES: NodeType> {
+    /// Stored results from computations
+    results: BTreeMap<TYPES::Epoch, DrbResult>,
+
+    /// Currently live computation
+    task: Option<(TYPES::Epoch, JoinHandle<DrbResult>)>,
+
+    /// Stored inputs to computations
+    seeds: BTreeMap<TYPES::Epoch, DrbSeedInput>,
+}
+
+impl<TYPES: NodeType> DrbComputations<TYPES> {
+    #[must_use]
+    /// Create a new DrbComputations
+    pub fn new() -> Self {
+        Self {
+            results: BTreeMap::new(),
+            task: None,
+            seeds: BTreeMap::new(),
+        }
+    }
+
+    /// If a task is currently live AND has finished, join it and save the result.
+    /// If the epoch for the calculation was the same as the provided epoch, return true
+    /// If a task is currently live and NOT finished, abort it UNLESS the task epoch is the same as
+    /// cur_epoch, in which case keep letting it run and return true.
+    /// Return false if a task should be spawned for the given epoch.
+    async fn join_or_abort_old_task(&mut self, epoch: TYPES::Epoch) -> bool {
+        if let Some((task_epoch, join_handle)) = &mut self.task {
+            if join_handle.is_finished() {
+                match join_handle.await {
+                    Ok(result) => {
+                        self.results.insert(*task_epoch, result);
+                        let result = *task_epoch == epoch;
+                        self.task = None;
+                        result
+                    }
+                    Err(e) => {
+                        tracing::error!("error joining DRB computation task: {e:?}");
+                        false
+                    }
+                }
+            } else if *task_epoch == epoch {
+                true
+            } else {
+                join_handle.abort();
+                self.task = None;
+                false
+            }
+        } else {
+            false
+        }
+    }
+
+    /// Stores a seed for a particular epoch for later use by start_new_task()
```

Comment:
Nit: Looks like `start_new_task` should be `handle_quorum_proposal_validated_drb_calculation_seed` now.